### PR TITLE
vtrace

### PR DESCRIPTION
**Pull request by le-horizon**
Vtrace implementation and tests.
Also added some documentation around TrainingInfo and the case of missing one TimeStep in on-policy training.

Most of the code is just for tests, and scalable_agent code copied over to be compared against.
Tests compare on policy case against existing alf discounted return and GAE, and off policy case with scalable_agent implementation.  There are also some tabular calculations.

Performance-wise, vtrace is slightly less stable than on policy AC.  However, when I lower the learning rate, vtrace is actually more sample efficient.  (AC with the high learning rate also crashes after an hour or so.)

Due to missing one timestep, I increased unroll length from 8 to 21, and plotted the curve.  21 is better.
This is the plot for breakout:
image

Similar trend is observed for dmlab explore goal small task, vtrace gets to return 50@100M steps, while ac only gets to 35.

In all these experiments, vtrace has 2 actors and thus twice as many environments than ac.

Thanks,
Le

### CODE REVIEW

File: alf/utils/value_ops.py [hunk 1]
```diff
@@ -286,3 +294,129 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+        return (1 - is_last) * (td + weighted_discount * vs_target_minus_vs)
+
+    vs_target_minus_vs = tf.scan(
+        fn=vs_target_minus_vs_fn,
+        elems=(weighted_discounts, tds, is_lasts),
+        initializer=tf.zeros_like(final_value),
+        reverse=True,
+        back_prop=False)
+
+    returns = (1 - is_lasts) * vs_target_minus_vs + values
```
Comment (emailweixu) [inline alf/utils/value_ops.py]: Seems the multiplication of `(1-is-lasts)` is unnecessary. It is already handled at line 346.

File: alf/examples/ac_simple_navigation.gin [hunk 2]
```diff
@@ -38,6 +38,11 @@ ActorCriticAlgorithm.value_network=@value/ValueNetwork()
 Agent.optimizer=@ac/Adam()
 Agent.gradient_clipping=None
 
+#import alf.algorithms.trac_algorithm
```
Comment (emailweixu) [inline alf/examples/ac_simple_navigation.gin]: do you intend to change this file?

File: README.md [hunk 3]
```diff
@@ -13,6 +13,7 @@ Agent Learning Framework (ALF) is a reinforcement learning framework emphasizing
 * [DDPG](alf/algorithms/ddpg_algorithm.py): Lillicrap et al. "Continuous control with deep reinforcement learning" [arXiv:1509.02971](https://arxiv.org/abs/1509.02971)
 * [PPO](alf/algorithms/ppo_algorithm.py): Schulman et al. "Proximal Policy Optimization Algorithms" [arXiv:1707.06347](https://arxiv.org/abs/1707.06347)
 * [SAC](alf/algorithms/sac_algorithm.py): Haarnoja et al. "Soft Actor-Critic Algorithms and Applications" [arXiv:1812.05905](https://arxiv.org/abs/1812.05905)
+* [VTrace/IMPALA]: Espeholt et al. "IMPALA: Scalable Distributed Deep-RL with Importance Weighted Actor-Learner Architectures" [arXiv:1802.01561v3]
```
Comment (emailweixu) [inline README.md]: link to source

File: alf/data_structures.py [hunk 4]
```diff
@@ -92,19 +92,32 @@ def make_action_time_step(time_step: TimeStep, prev_action, first_env_id=0):
         env_id=first_env_id + tf.range(time_step.step_type.shape[0]))
 
 
+# TrainingInfo contains data used for calculating losses during training.
+# It is a set of nested tensors of shape [B, T, ...], where B is batch size
+# and T is mini batch length, which equals the number of environment steps
+# to unroll (unroll_length) for on policy algorithms like actor critic or
+# off policy algorithms like PPO or V-Trace.
+#
+# The fields are populated in sync_off_policy_driver around line 143.
```
Comment (emailweixu) [inline alf/data_structures.py]: It's also populated in on_policy_driver

**Comment by le-horizon** [comment]
Done both.  Please take another look @emailweixu .

Thanks,
Le

File: README.md [hunk 5]
```diff
@@ -13,7 +13,7 @@ Agent Learning Framework (ALF) is a reinforcement learning framework emphasizing
 * [DDPG](alf/algorithms/ddpg_algorithm.py): Lillicrap et al. "Continuous control with deep reinforcement learning" [arXiv:1509.02971](https://arxiv.org/abs/1509.02971)
 * [PPO](alf/algorithms/ppo_algorithm.py): Schulman et al. "Proximal Policy Optimization Algorithms" [arXiv:1707.06347](https://arxiv.org/abs/1707.06347)
 * [SAC](alf/algorithms/sac_algorithm.py): Haarnoja et al. "Soft Actor-Critic Algorithms and Applications" [arXiv:1812.05905](https://arxiv.org/abs/1812.05905)
-* [VTrace/IMPALA]: Espeholt et al. "IMPALA: Scalable Distributed Deep-RL with Importance Weighted Actor-Learner Architectures" [arXiv:1802.01561v3]
+* [VTrace/IMPALA]: Espeholt et al. "IMPALA: Scalable Distributed Deep-RL with Importance Weighted Actor-Learner Architectures" [arXiv:1802.01561v3](https://arxiv.org/abs/1802.01561v3)
```
Comment (emailweixu) [inline README.md]: link to code.

File: README.md [hunk 6]
```diff
@@ -62,6 +63,15 @@ python -m alf.bin.play --root_dir=LOG_DIR
 
   <img src="alf/examples/ac_breakout.png" width="300" height="200" alt="breakout-training-curve"/> <img src="alf/examples/ac_breakout.gif" width="150" height="200" alt="breakout-playing-screen"/>
 
+For Vtrace/IMPALA with two actors and one learner (double the total number of parallel
```
Comment (emailweixu) [inline README.md]: a link to the gin file?

**Comment by le-horizon** [comment]
Added links to source code.  PTAL, thanks.

**Pull request merged**